function-controller/pkg/controller/autoscaler/autoscaler.go

```diff
@@ -0,0 +1,290 @@
+}
+
+func (a *autoScaler) emptyQueue(funcId FunctionId) (bool, int64) {
+	for topic, funcTotals := range a.totals {
+		if _, ok := funcTotals[funcId]; ok {
+			queueLen, err := a.transportInspector.QueueLength(topic, funcId.Function)
+			if err != nil {
+				log.Printf("Failed to obtain queue length (and will assume it is positive): %v", err)
+				return false, -1
+			}
+			if queueLen > 0 {
+				return false, queueLen
+			}
+		}
+	}
+	return true, 0
+}
+
+func (a *autoScaler) StartMonitoring(topic string, function FunctionId) error {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+	funcTotals, ok := a.totals[topic]
+	if !ok {
+		funcTotals = make(map[FunctionId]*metricsTotals)
+		a.totals[topic] = funcTotals
+	}
+
+	_, ok = funcTotals[function]
+	if ok {
+		return fmt.Errorf("Already monitoring topic %s and function %s", topic, function)
+	}
+
+	funcTotals[function] = &metricsTotals{}
+
+	a.proposals[function] = NewProposal(func() time.Duration {
+		return a.delayScaleDown(function.Function);
+	})
+
+	return nil
+}
+
+func (a *autoScaler) StopMonitoring(topic string, function FunctionId) error {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+	funcTotals, ok := a.totals[topic]
+	if !ok {
+		return fmt.Errorf("Not monitoring topic %s and function %s", topic, function)
+	}
+
+	_, ok = funcTotals[function]
+	if !ok {
+		return fmt.Errorf("Not monitoring topic %s and function %s", topic, function)
+	}
+
+	delete(funcTotals, function)
+
+	// Avoid leaking memory.
+	if len(funcTotals) == 0 {
+		delete(a.totals, topic)
+	}
+	delete(a.proposals, function)
+
+	return nil
+}
+
+func (a *autoScaler) receiveLoop() {
+	producerMetrics := a.metricsReceiver.ProducerMetrics()
+	consumerMetrics := a.metricsReceiver.ConsumerMetrics()
+	for {
+		select {
+		case pm, ok := <-producerMetrics:
+			if ok {
+				a.receiveProducerMetric(pm)
+			}
+
+		case cm, ok := <-consumerMetrics:
+			if ok {
+				a.receiveConsumerMetric(cm)
+			}
+
+		case <-a.stop:
+			close(a.accumulatingStopped)
+			return
+		}
+	}
+}
+
+func (a *autoScaler) receiveConsumerMetric(cm metrics.ConsumerAggregateMetric) {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+
+	funcTotals, ok := a.totals[cm.Topic]
+	if ok {
+		mt, ok := funcTotals[FunctionId{cm.Function}]
+		if ok {
+			mt.receiveCount += cm.Count
+		}
+	}
+}
+
+func (a *autoScaler) receiveProducerMetric(pm metrics.ProducerAggregateMetric) {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+
+	funcTotals, ok := a.totals[pm.Topic]
+	if ok {
+		for _, mt := range funcTotals {
+			mt.transmitCount += pm.Count
+		}
+	}
+}
+
+func (a *autoScaler) calculateProposal() {
+	for topic, funcTotals := range a.totals {
+		for fn, mt := range funcTotals {
+			var proposedReplicas int
+			if mt.receiveCount == 0 {
+				if mt.transmitCount == 0 {
+					proposedReplicas = 0
+				} else {
+					proposedReplicas = 1 // arbitrary value
+				}
+			} else {
+				proposedReplicas = int(math.Floor(float64(a.replicas[fn]) * float64(mt.transmitCount) / float64(mt.receiveCount)))
```

Comment:
Good point. Also we should try to round up rather than down as we don't want to let the input topic grow uncontrollably (i.e. at all, over time). I'll add a test and fix.

Reply:
Actually, it was a better point that I realised. The Google doc formula was wrong. It should actually be ceiling rather than floor. Fixed (with a test).